src/rust/cryptography-openssl/src/poly1305.rs

```diff
@@ -0,0 +1,48 @@
+// This file is dual licensed under the terms of the Apache License, Version
+// 2.0, and the BSD License. See the LICENSE file in the root of this repository
+// for complete details.
+
+#[cfg(any(CRYPTOGRAPHY_IS_BORINGSSL, CRYPTOGRAPHY_IS_LIBRESSL))]
+use std::mem::MaybeUninit;
+
+#[cfg(any(CRYPTOGRAPHY_IS_BORINGSSL, CRYPTOGRAPHY_IS_LIBRESSL))]
+pub struct Poly1305State {
+    // The state data must be allocated in the heap so that its address does not change. This is
+    // because BoringSSL APIs that take a `poly1305_state*` ignore all the data before an aligned
+    // address. Since a stack-allocated struct would change address on every copy, BoringSSL would
+    // interpret each copy differently, causing unexpected behavior.
+    context: Box<ffi::poly1305_state>,
+}
+
+#[cfg(any(CRYPTOGRAPHY_IS_BORINGSSL, CRYPTOGRAPHY_IS_LIBRESSL))]
+impl Poly1305State {
+    pub fn new(key: &[u8]) -> Poly1305State {
+        assert_eq!(key.len(), 32);
+        let mut ctx: Box<MaybeUninit<ffi::poly1305_state>> =
+            Box::new(MaybeUninit::<ffi::poly1305_state>::uninit());
+
+        // After initializing the context, unwrap the Box<MaybeUninit<poly1305_state>> into
+        // a Box<poly1305_state> while keeping the same memory address. See the docstring of the
+        // Poly1305State struct above for the rationale.
+        let initialized_ctx: Box<ffi::poly1305_state> = unsafe {
+            ffi::CRYPTO_poly1305_init(ctx.as_mut().as_mut_ptr(), key.as_ptr());
+            let raw_ctx_ptr = (*Box::into_raw(ctx)).as_mut_ptr();
```

Comment:
What is the deref + `as_mut_ptr` for?

Reply:
`into_raw()` consumes the `Box` and returns a `*mut MaybeUninit<poly1305_state>`. In order to go from `*mut MaybeUninit<poly1305_state>` to `*mut poly1305_state`, I need to call `MaybeUninit::as_mut_ptr` (which I can only do after I dereference). Unrolling the code above and adding type annotations, you get:
```rust
let raw_maybe_ptr: *mut MaybeUninit<ffi::poly1305_state> = Box::into_raw(ctx);
let raw_ctx_ptr: *mut ffi::poly1305_state = (*raw_maybe_ptr).as_mut_ptr();
Box::from_raw(raw_ctx_ptr)
```

Is there any other way of achieving this? There's [an API](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.assume_init) to convert from `Box<MaybeUninit<T>>` to `Box<T>`, but it's still experimental.

Reply:
Got it